Adjust arguments to Unsafe array operations while offheap is enabled

midronij opened the pull request:
In this contribution, the following adjustments have been made to the arguments that are passed in to Unsafe.get(), Unsafe.put(), and Unsafe.compareAndSwap() when offheap/balanced GC policy is in use:

- The dataAddr pointer is loaded and passed in as the base address of the object being operated on
- ~The array header size is subtracted from the offset~ 
_no longer necessary due to this change_: https://github.com/eclipse-openj9/openj9/pull/19326/

This is to account for the differing array shape under offheap/balanced GC policy as opposed to the default (gencon).

In addition, as a performance improvement for both the offheap and gencon cases, when the object type is known at compile time, we can eliminate unnecessary lowtag, class, and array checks that are currently performed at runtime for the following cases:

-  Object is known to be an array at compile time
- Object type is known to be non-array, non-class, non-java/lang/Object at compile time

Depends on https://github.com/eclipse-openj9/openj9/pull/18303

midronij commented:
@zl-wang @VermaSh @vijaysun-omr @0xdaryl here is the fix for the UnsafeAPI issue!

VermaSh commented:
@midronij Thank you for working on this! 
Did you test with latest Off heap changes [1-2]?

Also, can you please add a commit to remove checks which were added to disable unsafe array operations for off heap. You can find those checks in [1-2].



[1] [openj9/offHeapMemoryCommitAndDecommitFinal_combined_without_new_macro](https://github.com/VermaSh/openj9/tree/offHeapMemoryCommitAndDecommitFinal_combined_without_new_macro)
[2] [omr/offHeapMemoryCommitAndDecommitFinal_combined_without_new_macro](https://github.com/VermaSh/omr/tree/offHeapMemoryCommitAndDecommitFinal_combined_without_new_macro)

VermaSh commented:
Does this fix `unsafe.setMemory` as well?

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1231,18 +1248,49 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
       comp()->getDebug()->print(comp()->getOutFile(), lowTagCmpTree);
       }
 
+#if defined(TR_TARGET_64BIT)
```
For testing this is fine, but you'll need to replace that with `J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION`

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1369,22 +1417,85 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
+   TR::TreeTop *arrayAccessTreeTop = NULL;
+   TR::TreeTop *nonArrayAccessTreeTop = NULL;
+
+#if defined (TR_TARGET_64BIT)
```
Same here, you'll need to replace it with the new macro

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -884,11 +884,14 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
-   //if conversionNeeded is true, we haven't generated and we don't need arrayDirectAccessBlock
-   TR::Block *arrayDirectAccessBlock = conversionNeeded ? nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock() : NULL;
+
+   bool arrayBlockNeeded = conversionNeeded || (comp()->fej9()->isOffHeapAllocationEnabled() && comp()->target().is64Bit());
```
This was (or will be as part of newer vm/gc chagnes) moved to object model. You can use `TR::Compiler->om.isOffHeapAllocationEnabled()` instead but you don't need to guard with any macro as it will just return false in all other cases.

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1207,9 +1218,15 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+
+   if (conversionNeeded)
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(unsafeNodeWithConversion, conversionNeeded, false);
+   else if (comp()->fej9()->isOffHeapAllocationEnabled() && comp()->target().is64Bit())
```
Same as above about using object model instead of front end

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1231,18 +1248,49 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   //       lload      -> offset
+   //     iload        -> value to put
+
+   if (comp()->fej9()->isOffHeapAllocationEnabled())
```
Same as above about using object model instead of front end

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1369,22 +1417,85 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
+   TR::TreeTop *nonArrayAccessTreeTop = NULL;
+
+#if defined (TR_TARGET_64BIT)
+   if (comp()->fej9()->isOffHeapAllocationEnabled())
```
Same as above about using object model instead of front end

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1545,6 +1662,63 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
+
+   if (conversionNeeded)
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(callNodeWithConversion, conversionNeeded, true);
+   else if (comp()->fej9()->isOffHeapAllocationEnabled() && comp()->target().is64Bit())
```
Same as above about using object model instead of front end

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1545,6 +1662,63 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   TR::TreeTop* indirectAccessTreeTop =
+      genIndirectAccessCodeForUnsafeGetPut(callNodeTreeTop->getNode(), unsafeAddress);
+
+#if defined(TR_TARGET_64BIT)
```
Same thing here about the macro

VermaSh commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1545,6 +1662,63 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   //         aload    -> object base address
+   //         lload    -> offset
+
+   if (comp()->fej9()->isOffHeapAllocationEnabled())
```
Same as above about using object model instead of front end

VermaSh reviewed (commented):
Just a few minor suggestions. Also, here's the GC/VM PR these changes depend on https://github.com/eclipse-openj9/openj9/pull/14667

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1231,18 +1248,49 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+
+      //subtract header size from offset
+      TR::Node *oldOffset = addrToAccessNode->getChild(1);
+      TR::Node *adjustedOffset = TR::Node::create(TR::ladd, 2, oldOffset, TR::Node::lconst(-TR::Compiler->om.contiguousArrayHeaderSizeInBytes()));
```
if oldOffset is a constant to begin with, you don't need to gen ladd but put in a new constant node with the adjusted value.

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1369,22 +1417,85 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
+
+      //subtract header size from offset
+      TR::Node *oldOffset = CASicallNode->getChild(2);
+      TR::Node *adjustedOffset = TR::Node::create(TR::ladd, 2, oldOffset, TR::Node::lconst(-TR::Compiler->om.contiguousArrayHeaderSizeInBytes()));
```
if oldOffset is a constant to begin with, you don't need to gen ladd but put in a new constant node with the adjusted value.

zl-wang reviewed (commented):
i will read more to comment further next week

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -884,11 +884,14 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
    TR::TreeTop* nullComparisonTree = nullComparisonBlock->getLastRealTreeTop();
    TR::TreeTop *nullComparisonEntryTree = nullComparisonBlock->getEntry();
    TR::TreeTop *nullComparisonExitTree = nullComparisonBlock->getExit();
```
high level:  blocks layout is considered already (normal object is the most likely case)?  i.e.  directAccess should be the first block, then arrayAccess, then indirectAccess.

midronij commented:
@zl-wang I have finished addressing your review suggestions. As well, I have provided some examples of what the post-inlining IL trees will look like with my changes [here](https://github.com/eclipse-openj9/openj9/files/12525626/offheap_unsafe_examples.txt). Please let me know if there is anything else I should address with these additions.

Assuming everything looks good, I believe we are waiting on the following before we can merge this PR:
- https://github.com/eclipse-openj9/openj9/pull/14667 needs to get merged, since I believe this is where the `J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION` macro and the `TR::Compiler->om.isOffHeapAllocationEnabled()` check get implemented.
- As of right now, there appears to be a problem with one of the JIT helper methods, `getIntFromObject()`, which calls `Unsafe.getInt()`, which causes a NullPointerException when trying to build the git tests on vmfarm, and intermittently causes segmentation faults when running `javac` locally. I spoke to @VermaSh about this, and it seems that this was a previously existing issue which we are still waiting on a fix for. 
- If we want to include them in this PR, then `Unsafe.getAndAdd()`, `Unsafe.getAndSet()`, `Unsafe.setMemory()`, and `Unsafe.copyMemory()` also need to have their arguments adjusted to account for offheap array allocation. This is what I'm working on right now.

If there is anything that I've missed or that I'm mistaken about, please let me know!

zl-wang commented:
Works for the other Unsafe methods (if they are inlined in gencon ... this is certainly the condition) had better put up in a separate PR.

I supposed merging this PR has no effect on running none-offheap policies.  i.e.  it is relatively safe to merge.  every existing logics (for none-offheap) are still honoured.

Comments on the first example IL:
**BBStart <block_12>**  **ificmpeq** (equal to 0x10000) can be replaced with **ificmpne** (not equal to 0).  better code can be generated for the latter case, since constant 0x10000 is big for Power ISA;

zl-wang commented:
for the other two examples, i am wondering why we still generated the lowTag and class.class tests?  are they necessary? at least, you should check against the ILs generated for gencon

zl-wang commented:
the better question is:  if there is an example without any checks?  there must be

midronij commented:
> the better question is: if there is an example without any checks? there must be

For `Unsafe.get()` and `Unsafe.put()`, as far as I can tell, when using gencon GC policy (even prior to my changes), there is no case where no checks are performed. Normally, there is a lowtag check and a class check. If the J9Class of `java/lang/Class` cannot be loaded, then there is an array check and a lowtag check. So basically the lowtag check and one other is always generated no matter what. There's a little more detail on which checks are performed and when here: https://github.com/eclipse-openj9/openj9/blob/d3388e1662d364173da4c5e01a3a5001627c7946/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L1032-L1102
I'm not sure if there's a particular reason for that, or if it's simply because prior to my changes, we never actually checked if the object class was known at compile time.

midronij commented:
@zl-wang @VermaSh I've added the commit that gets rid of the uncessary lowtag/class/array checks when the object type is known at compile time. Once the necessary GC/VM changes are in this should be good to go

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1532,33 +1847,95 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   // Thus, in this scenario, rather than generate an additional block as the arrayDirectAccessBlock, we will just directly modify
+   // the directAccessBlock.
+   TR::TreeTop* arrayDirectAccessTreeTop = NULL;
+
```
There are 8 possible combinations among these 3 variables: conversionNeeded, arrayCheckNeeded, and arrayBlockNeeded. are you sure you covered them all with 3 prongs (if and else) below?

I guessed true-arrayCheckNeeded  implies true-arrayBlockNeeded, also true-conversionNeeded implies true-arrayBlockNeeded.  that reduced to 6 combinations.  Just point it out for you to double-check here if you got them all covered. At least, you can comment in code to that effect.

zl-wang replied:
when i read more of the code, the comment above looks like a general comment applicable to other places in code too.  you are doing if-then-else-if-then-else to generate different ILs.  it should be described to the effect that every possible case is covered.

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1120,6 +1191,36 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   bool needNotLowTagged = javaLangClass != NULL  || conversionNeeded ;
+
+   int length;
+   const char *objTypeSig = unsafeCall->getChild(1)->getSymbolReference()->getTypeSignature(length);
```
you need to consider the possible case where getChild(1) returns NULL.  i have seen such java code and dealt with such PMRs of crash.

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1120,6 +1191,36 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   // There are two cases where IL for type checks (array, class, lowtag) needs to be generated:
+   // 1.) Object type is unknown at compile time
+   // 2.) Object is known at compile time to be a java/lang/Class object (to match JNI implementation)
+   bool checksNeeded = objTypeSig == NULL || strstr(objTypeSig, "Ljava/lang/Object") || strstr(objTypeSig, "Ljava/lang/Class");
```
general comment on coding style:  use appropriate brackets for better readability.

zl-wang replied:
have you considered the possibility of **objTypeSig** being something like  **Ljava/lang/ObjectFoo** or **Ljava/lang/ClassBar**?  how about **objTypeSig** being **[Ljava/lang/Object**?

midronij commented:
@zl-wang changes for Unsafe.get/put/compareAndSwap are done! When you have a chance, could you please review and (if everything looks good) merge?

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1242,23 +1409,76 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+      TR::Node *oldOffset = addrToAccessNode->getChild(1);
+      TR::Node *adjustedOffset;
+
+      if (oldOffset->getOpCode().isLoadConst())
```
with expectation of ArrayBaseOffset being zero in off-heap case, why do we still need offset adjustment? this question is applicable to a few cases below.

midronij replied:
I think this must have occurred due to some issues with conflict resolution when I was cleaning up the commits in this branch. I'll fix that and make sure that the rest of the code for this contribution is up-to-date as well.

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -895,11 +897,20 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
+
+   // There are 5 possible cases determining which blocks should be added to the generated IL trees
+   // 1.) checksNeeded && arrayCheckNeeded && arrayBlockNeeded: add directAccess, indirectAccess, and arrayDirectAccess
+   // 2.) checksNeeded && arrayCheckNeeded && !arrayBlockNeeded: add directAccess and indirectAccess only
```
which scenario is case2? some comments are due here. i meant: if arrayBlockNeeded is false, how can arrayCheckNeeded be true? what is the arrayChecking for? i.e. if it is an array (the checking result), you still don't need arrayBlock. why bother to check it to begin with?

midronij replied:
I believe this is a holdover from before we decided to check if the object class is known at compile time (i.e.: all type checks were performed at runtime no matter what). This case would only occur with when using gencon GC policy, if the J9Class of the object could not be retrieved (i.e.: `javaLangClass == NULL`). But you're right that the array check shouldn't be necessary in the first place since the only condition that changes whether a direct/indirect access should be used is if the object type is `java/lang/Class`.

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -895,11 +897,20 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
+   TR::Block *indirectAccessBlock;
+
+   // There are 5 possible cases determining which blocks should be added to the generated IL trees
+   // 1.) checksNeeded && arrayCheckNeeded && arrayBlockNeeded: add directAccess, indirectAccess, and arrayDirectAccess
```
for non-offheap policy (i.e.  normal object and array headers are of same size), we can optimize the cases down to the fact that we don't need arrayCheck, since object and array accesses can use the same tree. i presumed you took advantage of that optimization in higher level of routines where you can always set arrayCheckNeeded to false (for gencon e.g.).

midronij replied:
In the case where the data being read/written is less than 4 bytes, object and array accesses are actually handled differently, as outlined [here](https://github.com/eclipse-openj9/openj9/blob/671e76243d8e011486e26f4d2d298012efe2e2d9/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L1041-L1111): 
```
Converting Unsafe.get/put* routines into inlined code involves two cases:
1) if the size of the element to put/get is 4 bytes or more
2) if the size of the element to put/get is less than 4 bytes (boolean, byte, char, short)
.
.
.
In (2), there are three alternatives on how to read from/write the object. direct,
direct with conversion, indirect. The same three conditions are used to decide which one
to use based on the following pseudocode:
if (object is NULL)
  use direct access with conversion
else if (object is array)
  use direct access with conversion
else if (object is of type Class)
  use indirect access
else
  use direct access
```
...which is why I still included the array check in the case where gencon GC policy is being used and the object type is unknown at compile time.

zl-wang reviewed (changes requested):
High level comment: as it is, the implementation is fairly complicated.  If you can simplify it, that would be good.

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1181,6 +1248,43 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   // 1.) Object type info is unknown/inaccessible
+   // 2.) Object type is known at compile time to be java/lang/Object
+   // 3.) Object type is known at compile time to be one of the interface classes java/lang/Cloneable or java/io/Serializable
+   bool objTypeUnknown = objTypeSig == NULL || strncmp(objTypeSig, "Ljava/lang/Object;", strlen("Ljava/lang/Object;")) == 0
```
check out abdul's latest work on this: it is JLO (java/lang/Object) or any interface type. change correspondingly.

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1399,31 +1547,133 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
+   // 1.) Object type info is unknown/inaccessible
+   // 2.) Object type is known at compile time to be java/lang/Object
+   // 3.) Object type is known at compile time to be one of the interface classes java/lang/Cloneable or java/io/Serializable
+   bool objTypeUnknown = objTypeSig == NULL || strncmp(objTypeSig, "Ljava/lang/Object;", strlen("Ljava/lang/Object;")) == 0
```
Ditto.

midronij commented:
@zl-wang I have attached the trace logs for every possible scenario, for both gencon and offheap, as well as a summary of what the IL trees should look like after inlining for each case. If there is any other information I should provide please let me know

[unsafe_gencon.log](https://github.com/user-attachments/files/16090512/unsafe_gencon.log)
[unsafe_offheap.log](https://github.com/user-attachments/files/16090513/unsafe_offheap.log)
[logs.tar.gz](https://github.com/user-attachments/files/16090510/logs.tar.gz)

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -831,11 +831,14 @@ TR_J9InlinerPolicy::createAnchorNodesForUnsafeGetPut(TR::TreeTop* treeTop,
-      compRefTT->join(isUnsafeGet ? treeTop : treeTop->getNextTreeTop());
+         TR::TreeTop::create(comp(), genCompressedRefs(isUnsafeGet?node->getFirstChild():node,
+                                                      false));
+      if (compRefTT)
```
do you ever expect compRefTT can be NULL? it came from `create()` ... if it failed to create, compilation was expected to be aborted before return, I believed.

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1181,6 +1248,51 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   // There are two cases where IL for type checks (array, class, lowtag) needs to be generated:
+   // 1.) Object type is unknown at compile time
+   // 2.) Object is known at compile time to be a java/lang/Class object (to match JNI implementation)
+   bool checksNeeded = objTypeUnknown || (strncmp(objTypeSig, "Ljava/lang/Class;", strlen("Ljava/lang/Class;")) == 0);
```
doesn't need this `strncmp` anymore, since `objClass == comp()->getObjectClassPointer()` is exactly doing that already.  you can find out what is `getObjectClassPointer()` returns from source code:  it is exactly JLO.

midronij replied:
I believe this `strncmp` is for java/lang/Class, not java/lang/Object. Since this a special case (i.e.: requires indirect access rather than direct access), it has to be included in addition to the `objTypeUnknown` condition (which includes JLO and interface types).

That being said, since we're no longer using `strncmp` like this for JLO or the interface types, I can do the same for java/lang/Class as well, if that would be preferable.

zl-wang replied:
ah, noticed the difference.  a further question is:  you didn't do this java/lang/Class test in CAS case.  is that expected?

midronij replied:
You're right, I didn't do that, but that possibility definitely needs to be accounted for. I'll add that condition to the CAS case now. Thank you for catching that!

zl-wang commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1494,6 +1783,51 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   // There are two cases where IL for type checks (array, class, lowtag) needs to be generated:
+   // 1.) Object type is unknown at compile time
+   // 2.) Object is known at compile time to be a java/lang/Class object (to match JNI implementation)
+   bool checksNeeded = objTypeUnknown || (strncmp(objTypeSig, "Ljava/lang/Class;", strlen("Ljava/lang/Class;")) == 0);
```
Ditto.

hzongaro commented:
> I have attached the trace logs for every possible scenario, for both gencon and offheap, as well as a summary of what the IL trees should look like after inlining for each case.

@midronij, in the cases that generate run-time comparisons to the j9Class for `java/lang/Class`, there are variations that will be generated if `comp()->getClassClassPointer()` returns a null pointer.  May I ask you to show those variations?  (It should be possible to force that by replacing various calls to `comp()->getClassClassPointer()` with `NULL`.

hzongaro reviewed (commented):
I haven't finished reviewing, but I thought I would provide my comments thus far, most of which are minor.

One thing I wanted to point out in looking at the complete log files in [logs.tar.gz](https://github.com/user-attachments/files/16090510/logs.tar.gz) is that verification errors occurred after inlining in the off-heap cases.  For instance, in `offheap_array.log`, the following lines appear:

```
TREE VERIFICATION ERROR -- node [    0x7a5d54970820] ref count is 0 and should be 1
BLOCK VERIFICATION ERROR -- node [    0x7a5d54970820] accessed outside of its (extended) basic block: -1 time(s)
```

I'm not sure whether those were related to the stray call to `genClassCheckForUnsafeGetPut` that I pointed out, or if there is some other place that's generated IL that was misplaced.  I thought it would be better to let you know now, and then I'll resume reviewing once that error has been fixed.

midronij commented:
@hzongaro thank you for your feedback! I have fixed the tree verification errors and addressed your other comments, so whenever you have a chance please feel free to continue reviewing.

Regarding the case where `comp()->getClassClassPointer()` returns `NULL`; I believe that case is covered by checking the condition `javaLangClass == NULL`, like so: 
```
TR_OpaqueClassBlock *javaLangClass = comp()->getClassClassPointer(/* isVettedForAOT = */ true);
.
.
.
// We need to generate arrayDirectAccessBlock if BOTH of following conditions hold:
// - conversionNeeded == true OR javaLangClass == NULL OR offheap is enabled
// - object is known be an array at compile time OR object type is unknown at compile time
bool arrayBlockNeeded = (conversionNeeded || javaLangClass == NULL || TR::Compiler->om.isOffHeapAllocationEnabled() && comp()->target().is64Bit()) &&
                           (objTypeUnknown || objTypeSig[0] == '[');

// We need to generate IL for the array check if we have determined that IL type checks are needed (i.e.: checksNeeded == true) AND
// a separate block is needed to handle array access (i.e.: arrayBlockNeeded == true) AND at least one of the following conditions holds:
// - conversionNeeded == true
// - javaLangClass == NULL
// - offheap is enabled
bool arrayCheckNeeded = checksNeeded && arrayBlockNeeded && (conversionNeeded || javaLangClass == NULL || (TR::Compiler->om.isOffHeapAllocationEnabled() && comp()->target().is64Bit()));
```
(full code can be viewed [here](https://github.com/midronij/openj9/blob/8f53deab8e36ed247a67a6abf231584b1e039946/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L1256-L1294))

So for example, in the scenario that you mentioned above, where gencon GC policy is being used, `conversionNeeded == false` (i.e.: `getInt()` or `getLong()`) and the type of the object is unknown at compile time: the `NULL`, array, and lowtag runtime checks are generated. As far as I know this should be consistent with how this scenario was dealt with before my changes came into the picture.

You can take a look at the full trace log for that scenario here: 
[gencon_int_unknown_javalangclassNULL.log](https://github.com/user-attachments/files/16133509/gencon_int_unknown_javalangclassNULL.log)

hzongaro commented:
> As far as I know this should be consistent with how this scenario was dealt with before my changes came into the picture.

Thanks!  I saw that the code still handled that case after your changes - I just wanted to make sure you were able to exercise it generating a log file.

hzongaro commented on runtime/compiler/optimizer/J9Inliner.hpp:
```diff
@@ -313,14 +313,21 @@ class TR_J9InlinerPolicy : public OMR_InlinerPolicy
        * \param conversionNeeded Indicates whether the call reprents an \c Unsafe
```
It looks like you accidentally undid this change.

midronij replied:
Seems like I somehow dropped the commit that included that change. My mistake! It's been added back now.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1181,6 +1244,51 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   // So in that case we will generate lowTagCmpTest to branch to indirect access if true
+   TR_OpaqueClassBlock *javaLangClass = comp()->getClassClassPointer(/* isVettedForAOT = */ true);
+
+   int length;
+   const char *objTypeSig = unsafeCall->getChild(1)->getSymbolReference() ? unsafeCall->getChild(1)->getSymbolReference()->getTypeSignature(length) : NULL;
+
+   // There are three cases where we cannot be sure of the Object type at compile time:
+   // 1.) Object type info is unknown/inaccessible
+   // 2.) Object type is known at compile time to be java/lang/Object
+   // 3.) Object type is known at compile time to be some interface class (e.g.: java/lang/Cloneable, java/io/Serializable)
+   bool objTypeUnknown;
+   TR_OpaqueClassBlock *objClass = NULL;
+
+   if (objTypeSig == NULL)
+      {
+      objTypeUnknown = true;
+      }
+   else
+      {
+      objClass = comp()->fe()->getClassFromSignature(objTypeSig, length, unsafeCall->getChild(1)->getSymbolReference()->getOwningMethod(comp()));
+      objTypeUnknown = objClass == NULL || objClass == comp()->getObjectClassPointer() || TR::Compiler->cls.isInterfaceClass(comp(), objClass);
+      }
```
I vaguely recall a discussion in which @jdmpapin mentioned that the type signature is not necessarily reliable for parameters, as the parameter slot might be reused for an object of a different type.  @jdmpapin, @vijaysun-omr, is there a safe way to query this information?  When can `getTypeSignature` be relied upon?

midronij replied:
I think this may be the discussion in question: https://github.com/eclipse-openj9/openj9/pull/19634#discussion_r1649400338.

jdmpapin replied:
Henry, that's right. This comment reminds me that I forgot about this problem when reviewing #19634, so the merged code [incorrectly trusts parameter signatures here](https://github.com/eclipse-openj9/openj9/blob/abc6fa4ae389c7fbe117d0be820eecdf70d30140/runtime/compiler/optimizer/J9TransformUtil.cpp#L2981-L3008). The comment just linked was about something else (the unreliability of interface types in signatures, and the fact that autos and array elements don't even have a signature)

In general `getTypeSignature()` can be trusted if it's called on anything but a parameter and the signature doesn't name an interface type (including as an array component type). If the signature does name an interface type, then we could get an accurate type bound by erasing the interface type to `Ljava/lang/Object;`. In the case of a parameter, the signature is accurate at method entry. It can potentially be relied upon if some additional analysis ensures that an incompatible type of reference in never stored into the parameter (or hasn't been in any definition that can reach a particular use). Such analysis could be imprecise, e.g. trust the signature if there are no stores to the parameter

vijaysun-omr replied:
Some of those cases are mentioned in the comment in the code snippet above. There may be other cases that @jdmpapin might mention.

zl-wang replied:
i.e. you need to test if `unsafecall->getChild(1)` is a parameter from the being-compiled method. if it is, treat it as if it were a JLO.

jdmpapin replied:
Also, a note about the code comment `3.)`. The code itself correctly distrusts any interface type, but the examples in the comment seem to me to suggest that interface types might be a concern only because some interfaces are actually implemented by array types

midronij replied:
In that case, would something like this account for the scenarios where `getTypeSignature()` can't be trusted?
```
bool objTypeUnknown;
.
.
.
bool isParameter = unsafeAddress->getSymbolReference()->getSymbol()->isParm();
objTypeUnknown = objClass == NULL || objClass == comp()->getObjectClassPointer() || 
                 TR::Compiler->cls.isInterfaceClass(comp(), objClass) || isParameter;

```
In regards to Devin's other comment, I can also update the comment for this condition so that it's more clear why interfaces are being included here:
```
// 3.) Object type is known at compile time to be some interface class (e.g.: java/lang/Cloneable, java/io/Serializable)
//     - Note that interfaces are included in this condition because valid bytecode can store an array into an interface type, 
//       which can then be passed in to Unsafe.get()
```

jdmpapin replied:
Yes, conservatively treating all parameters as unknown-type will work. (I assume that `unsafeAddress` here is `unsafeCall->getChild(1)`)

To me the proposed comment still doesn't quite explain why all interfaces are rejected. The same comment would also describe the situation where only `Cloneable` and `Serializable` interfere, but where we conservatively avoid them by avoiding all interfaces. I'd also say that "Object type is known at compile time to be some interface class" is misleading. In fact when the signature names an interface type, the object type is completely unknown at compile time (just as the name `objTypeUnknown` correctly suggests)

If I may, here's a suggestion:
```
// There are four cases where we cannot be sure of the Object type at compile time:
// 1.) No signature is available
// 2.) The signature names java/lang/Object
// 3.) The signature names an interface type, which is a problem because
//     interface types are not checked by the verifier
// 4.) The signature belongs to a parameter, which is a problem because
//     parameters can share slots with other variables
```

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -851,27 +853,53 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
                                        TR::TreeTop* indirectAccessTreeTop,
                                        bool needNullCheck, bool isUnsafeGet,
                                        bool conversionNeeded,
+                                       bool arrayCheckNeeded, bool arrayBlockNeeded,
+                                       bool checksNeeded,
                                        TR::Node* orderedCallNode = NULL)
    {
+   // There are 4 possible cases determining which checks and blocks should be added to the generated IL trees:
+   // 1.) checksNeeded && arrayCheckNeeded && arrayBlockNeeded
+   //     (i.e.: (offheap AND object type is unknown at compile time) OR conversionNeeded)
+   //     - checks: NULLCHK, ArrayCHK, LowtagCHK
+   //     - blocks: arrayDirectAccess, directAccess, indirectAccess
+   // 2.) checksNeeded && !arrayCheckNeeded && !arrayBlockNeeded
+   //     (i.e.: gencon AND object type is unknown at compile time AND !conversionNeeded)
```
It doesn't look like `conversionNeeded` is used any longer after the new `arrayCheckNeeded`, `arrayBlockNeeded` and `checksNeeded` are introduced.  If that's correct, the parameter should be removed and the comments adjusted.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -982,35 +1043,29 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
 
       TR::Node *testIsArrayFlag = comp()->fej9()->testIsClassArrayType(vftLoad);
       TR::Node *flagConstNode = TR::Node::create(testIsArrayFlag, TR::iconst, 0, TR::Compiler->cls.flagValueForArrayCheck(comp()));
-      TR::Node *isArrayNode = TR::Node::createif(TR::ificmpeq, testIsArrayFlag, flagConstNode, NULL);
+      TR::Node *isArrayNode = TR::Node::createif(TR::ificmpne, testIsArrayFlag, TR::Node::create(TR::iconst, 0), NULL);
```
The `flagConstNode` is no longer after changing to compare with the result of `testIsArrayFlag` with zero, so no need to generate it.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1059,12 +1114,15 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
    if (needNullCheck)
       {
-      TR::TreeTop *treeBeforeCmp = firstComparisonTree->getPrevTreeTop();
+      TR::TreeTop *treeBeforeCmp = firstComparisonTree ? firstComparisonTree->getPrevTreeTop() : beforeCallBlock->getExit()->getPrevTreeTop();
```
In case (4) the block won't be split and `firstComparisonTree` will be a null pointer.  Will this end up placing the `NULLCHK` after the point where the call originally appeared?

midronij replied:
In case 4, the `NULLCHK` ends up being placed like so: 
```
n1n       BBStart <block_2>                                                                   [    0x7b404f004660] bci=[-1,0,33] rc=0 vc=14 vn=- li=- udi=- nc=0
n42n      NULLCHK on n3n [#32]                                                                [    0x7b404f005330] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=1
n41n        PassThrough                                                                       [    0x7b404f0052e0] bci=[-1,5,33] rc=1 vc=0 vn=- li=- udi=- nc=1
n3n           aload  TestUnsafe.unsafe Lsun/misc/Unsafe;[#405  Static] [flags 0x307 0x0 ]     [    0x7b404f004700] bci=[-1,0,33] rc=1 vc=14 vn=- li=- udi=- nc=0
n44n      astore  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                                  [    0x7b404f0053d0] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=1
n4n         aload  <parm 0 [I>[#403  Parm] [flags 0x40000107 0x0 ]                            [    0x7b404f004750] bci=[-1,3,33] rc=1 vc=14 vn=- li=- udi=- nc=0
n46n      lstore  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                                  [    0x7b404f005470] bci=[-1,4,33] rc=0 vc=0 vn=- li=- udi=- nc=1
n5n         lload  <parm 1 J>[#404  Parm] [flags 0x40000104 0x0 ]                             [    0x7b404f0047a0] bci=[-1,4,33] rc=1 vc=14 vn=- li=- udi=- nc=0
n72n      BBEnd </block_2> =====                                                              [    0x7b404f0a0640] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
```
I believe that's the expected behaviour here...but of course, if I'm mistaken then please let me know and I'll fix it.

hzongaro replied:
Oh, I see.  I missed the fact that the block actually is being split in case (4) here:  https://github.com/eclipse-openj9/openj9/blob/b42906eb7859c07ba5782a02e8dd56dbec96aaad/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L946-L953

hzongaro reviewed (commented):
I still need to review the changes to the code that calls `genCodeForUnsafeGetPut`, but I wanted to pass along my review comments thus far.

midronij commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -851,27 +853,53 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
                                        TR::TreeTop* indirectAccessTreeTop,
                                        bool needNullCheck, bool isUnsafeGet,
                                        bool conversionNeeded,
+                                       bool arrayCheckNeeded, bool arrayBlockNeeded,
+                                       bool checksNeeded,
                                        TR::Node* orderedCallNode = NULL)
    {
+   // There are 4 possible cases determining which checks and blocks should be added to the generated IL trees:
+   // 1.) checksNeeded && arrayCheckNeeded && arrayBlockNeeded
+   //     (i.e.: (offheap AND object type is unknown at compile time) OR conversionNeeded)
+   //     - checks: NULLCHK, ArrayCHK, LowtagCHK
+   //     - blocks: arrayDirectAccess, directAccess, indirectAccess
+   // 2.) checksNeeded && !arrayCheckNeeded && !arrayBlockNeeded
+   //     (i.e.: gencon AND object type is unknown at compile time AND !conversionNeeded)
+   //     - checks: LowtagCHK, NULLCHK, ClassCHK
+   //     - blocks: directAccess, indirectAccess
+   // 3.) !checksNeeded && !arrayCheckNeeded && arrayBlockNeeded
+   //     (i.e.: offheap AND object is known to be array at compile time)
+   //     - checks: NULLCHK
+   //     - blocks: arrayDirectAccess, directAccess
+   // 4.) !checksNeeded && !arrayCheckNeeded && !arrayBlockNeeded
+   //     (i.e.: (offheap AND object is known to be non-array object at compile) time OR gencon)
```
I went with `arrayTestNeeded` and `typeTestsNeeded` to replace `arrayCheckNeeded` and `checksNeeded` for now. I am a little hesitant to rename `typeTestsNeeded` to `lowTagTestNeeded`, just because `typeTestsNeeded == true` is a requirement for `arrayTestNeeded` to evaluate to `true`. The intent there is that we first need to check if ANY tests at all (which could include a lowtag test, an array test, and/or a class test) are necessary, and _then_ determine if an array test specifically is needed. I worry that naming it `lowTagTestNeeded` may not get that across clearly. 

Same with the concern about the parameter names being too descriptive of the implementation details. I agree that it's not ideal, but I'm not sure that there's another way to communicate what those variables do efficiently...at least not with how they are currently used.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1270,9 +1379,20 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   if (conversionNeeded && (arrayTestNeeded|| arrayBlockNeeded))
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(unsafeNodeWithConversion, conversionNeeded, false);
+   else if (arrayTestNeeded || arrayBlockNeeded)
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(unsafeNode, false, false);
```
I just realized that in this case, if `javaLangClass == NULL`, but `conversionNeeded` is false and `TR::Compiler->om.isOffHeapAllocationEnabled()` is false, the trees that are generated for `arrayDirectAccessTreeTop` are identical to those generated for `directAccessTreetop` above.

I can see that that's necessary for the off-heap allocation case, as you need to tailor the IL under `arrayDirectAccessTreeTop` for the off-heap array access below, but otherwise, this results in duplicated blocks.  The duplicated block can be seen in the [log you generated earlier](https://github.com/user-attachments/files/16133509/gencon_int_unknown_javalangclassNULL.log) for the `javaLangClass == NULL` case &mdash; block_9 and block_8 are the same (except for a `goto` operation):

```
n105n     BBStart <block_12>                                                                  [    0x7c44f30a1090] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n104n     ificmpne --> block_8 BBStart at n86n ()                                             [    0x7c44f30a1040] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=2 flg=0x20
n101n       iand                                                                              [    0x7c44f30a0f50] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=2
n99n          l2i                                                                             [    0x7c44f30a0eb0] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=1
n98n            lloadi  <isClassDepthAndFlags>[#317  Shadow +24] [flags 0x603 0x0 ]           [    0x7c44f30a0e60] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=1
n95n              aloadi  <vft-symbol>[#320  Shadow] [flags 0x18607 0x0 ]                     [    0x7c44f30a0d70] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=1
n94n                aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                         [    0x7c44f30a0d20] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n100n         iconst 0x10000 (X!=0 X>=0 )                                                     [    0x7c44f30a0f00] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x104
n103n       iconst 0                                                                          [    0x7c44f30a0ff0] bci=[-1,0,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n106n     BBEnd </block_12> =====                                                             [    0x7c44f30a10e0] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=0

n96n      BBStart <block_11>                                                                  [    0x7c44f30a0dc0] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n83n      iflcmpne --> block_10 BBStart at n91n ()                                            [    0x7c44f30a09b0] bci=[-1,4,33] rc=0 vc=0 vn=- li=- udi=- nc=2 flg=0x20
n82n        land                                                                              [    0x7c44f30a0960] bci=[-1,4,33] rc=1 vc=0 vn=- li=- udi=- nc=2
n47n          lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                               [    0x7c44f30054c0] bci=[-1,4,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n81n          lconst 1 (highWordZero X!=0 X>=0 )                                              [    0x7c44f30a0910] bci=[-1,0,33] rc=2 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
n81n        ==>lconst 1
n97n      BBEnd </block_11> =====                                                             [    0x7c44f30a0e10] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=0

n89n      BBStart <block_9>                                                                   [    0x7c44f30a0b90] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n53n      istore  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ]                                  [    0x7c44f30a0050] bci=[-1,5,33] rc=0 vc=14 vn=- li=- udi=- nc=1
n54n        iloadi  <unsafe shadow sym>[#420  Shadow] [flags 0x80000603 0x100 ] ()            [    0x7c44f30a00a0] bci=[-1,5,33] rc=1 vc=15 vn=- li=- udi=- nc=1 flg=0x20
n55n          aladd                                                                           [    0x7c44f30a00f0] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=2
n56n            aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                             [    0x7c44f30a0140] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n58n            lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                             [    0x7c44f30a01e0] bci=[-1,4,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n90n      BBEnd </block_9> =====                                                              [    0x7c44f30a0be0] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0

n84n      BBStart <block_7>                                                                   [    0x7c44f30a0a00] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n8n       ireturn                                                                             [    0x7c44f3004890] bci=[-1,8,33] rc=0 vc=14 vn=- li=- udi=- nc=1
n6n         iload  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ] ()                              [    0x7c44f30047f0] bci=[-1,5,33] rc=1 vc=15 vn=- li=- udi=- nc=0 flg=0x20
n2n       BBEnd </block_7> =====                                                              [    0x7c44f30046b0] bci=[-1,8,33] rc=0 vc=14 vn=- li=- udi=- nc=0

n86n      BBStart <block_8>                                                                   [    0x7c44f30a0aa0] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n60n      istore  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ]                                  [    0x7c44f30a0280] bci=[-1,5,33] rc=0 vc=14 vn=- li=- udi=- nc=1
n61n        iloadi  <unsafe shadow sym>[#420  Shadow] [flags 0x80000603 0x100 ] ()            [    0x7c44f30a02d0] bci=[-1,5,33] rc=1 vc=15 vn=- li=- udi=- nc=1 flg=0x20
n62n          aladd                                                                           [    0x7c44f30a0320] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=2
n63n            aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                             [    0x7c44f30a0370] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n65n            lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                             [    0x7c44f30a0410] bci=[-1,4,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n88n      goto --> block_7 BBStart at n84n                                                    [    0x7c44f30a0b40] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n87n      BBEnd </block_8> =====                             
```

This might not be terribly important, as I hope the `javaLangClass == NULL` condition doesn't occur frequently, but it might be worth thinking about.

hzongaro replied:
Thinking about it a little more, I think this could be handled with `arrayTestNeeded` being set to true, but `arrayBlockNeeded` set to false.  `genCodeForUnsafeGetPut` has had to deal with that scenario in the past.

midronij replied:
Thank you for catching that! I've modified my changes to handle this situation as a special case, since as you said, it's not something we expected to come up too much. In the situation where gencon is being used and `conversionNeeded` is false, both sides of the array test diamond will lead to the `directAccessBlock`, so we can just skip the array test altogether. This leaves just the NULL and lowtag tests, like so:
```
n1n       BBStart <block_2>                                                                   [    0x71c90ec14660] bci=[-1,0,36] rc=0 vc=14 vn=- li=- udi=- nc=0
n42n      NULLCHK on n3n [#32]                                                                [    0x71c90ec15330] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=1
n41n        PassThrough                                                                       [    0x71c90ec152e0] bci=[-1,5,36] rc=1 vc=0 vn=- li=- udi=- nc=1
n3n           aload  TestUnsafe.unsafe Lsun/misc/Unsafe;[#405  Static] [flags 0x307 0x0 ]     [    0x71c90ec14700] bci=[-1,0,36] rc=1 vc=14 vn=- li=- udi=- nc=0
n44n      astore  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                                  [    0x71c90ec153d0] bci=[-1,3,36] rc=0 vc=0 vn=- li=- udi=- nc=1
n4n         aload  <parm 0 Ljava/lang/Object;>[#403  Parm] [flags 0x40000107 0x0 ]            [    0x71c90ec14750] bci=[-1,3,36] rc=1 vc=14 vn=- li=- udi=- nc=0
n46n      lstore  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                                  [    0x71c90ec15470] bci=[-1,4,36] rc=0 vc=0 vn=- li=- udi=- nc=1
n5n         lload  <parm 1 J>[#404  Parm] [flags 0x40000104 0x0 ]                             [    0x71c90ec147a0] bci=[-1,4,36] rc=1 vc=14 vn=- li=- udi=- nc=0
n73n      ifacmpeq --> block_9 BBStart at n82n ()                                             [    0x71c90ecb0690] bci=[-1,3,36] rc=0 vc=0 vn=- li=- udi=- nc=2 flg=0x20
n71n        aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                                 [    0x71c90ecb05f0] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n72n        aconst NULL (X==0 X>=0 X<=0 )                                                     [    0x71c90ecb0640] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x302
n78n      BBEnd </block_2> =====                                                              [    0x71c90ecb0820] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0

n86n      BBStart <block_10>                                                                  [    0x71c90ecb0aa0] bci=[-1,3,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n76n      iflcmpeq --> block_8 BBStart at n79n ()                                             [    0x71c90ecb0780] bci=[-1,4,36] rc=0 vc=0 vn=- li=- udi=- nc=2 flg=0x20
n75n        land                                                                              [    0x71c90ecb0730] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=2
n47n          lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                               [    0x71c90ec154c0] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n74n          lconst 1 (highWordZero X!=0 X>=0 )                                              [    0x71c90ecb06e0] bci=[-1,0,36] rc=2 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
n74n        ==>lconst 1
n87n      BBEnd </block_10> =====                                                             [    0x71c90ecb0af0] bci=[-1,3,36] rc=0 vc=0 vn=- li=- udi=- nc=0

n82n      BBStart <block_9>                                                                   [    0x71c90ecb0960] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n53n      istore  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ]                                  [    0x71c90ecb0050] bci=[-1,5,36] rc=0 vc=14 vn=- li=- udi=- nc=1
n54n        iloadi  <unsafe shadow sym>[#420  Shadow] [flags 0x80000603 0x100 ] ()            [    0x71c90ecb00a0] bci=[-1,5,36] rc=1 vc=15 vn=- li=- udi=- nc=1 flg=0x20
n55n          aladd                                                                           [    0x71c90ecb00f0] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=2
n56n            aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                             [    0x71c90ecb0140] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n58n            lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                             [    0x71c90ecb01e0] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n83n      BBEnd </block_9> =====                                                              [    0x71c90ecb09b0] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0

n77n      BBStart <block_7>                                                                   [    0x71c90ecb07d0] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n8n       ireturn                                                                             [    0x71c90ec14890] bci=[-1,8,36] rc=0 vc=14 vn=- li=- udi=- nc=1
n6n         iload  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ] ()                              [    0x71c90ec147f0] bci=[-1,5,36] rc=1 vc=15 vn=- li=- udi=- nc=0 flg=0x20
n2n       BBEnd </block_7> =====                                                              [    0x71c90ec146b0] bci=[-1,8,36] rc=0 vc=14 vn=- li=- udi=- nc=0

n79n      BBStart <block_8> (freq 1) (cold)                                                   [    0x71c90ecb0870] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n62n      istore  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ]                                  [    0x71c90ecb0320] bci=[-1,5,36] rc=0 vc=14 vn=- li=- udi=- nc=1
n63n        iloadi  <unsafe shadow sym>[#422  Shadow] [flags 0x80000603 0x100 ] ()            [    0x71c90ecb0370] bci=[-1,5,36] rc=1 vc=15 vn=- li=- udi=- nc=1 flg=0x20
n64n          aladd                                                                           [    0x71c90ecb03c0] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=2
n65n            aloadi  <ramStaticsFromClass>[#313  Shadow +72] [flags 0x10607 0x0 ]          [    0x71c90ecb0410] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=1
n66n              aloadi  <classFromJavaLangClass>[#311  Shadow +8] [flags 0x10607 0x0 ]      [    0x71c90ecb0460] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=1
n67n                aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                         [    0x71c90ecb04b0] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n68n            land                                                                          [    0x71c90ecb0500] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=2
n69n              lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                           [    0x71c90ecb0550] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n70n              lconst -4 (X!=0 X<=0 )                                                      [    0x71c90ecb05a0] bci=[-1,5,36] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x204
n81n      goto --> block_7 BBStart at n77n                                                    [    0x71c90ecb0910] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n80n      BBEnd </block_8> (cold) =====  
```
(The full trace log can be viewed [here](https://github.com/user-attachments/files/16240212/gencon_int_unknown_javaLangClassNULL.log).)

In adding this additional case, I also found that I was able to simplify the logic surrounding `typeTestsNeeded`, `arrayBlockNeeded`, and `arrayTestNeeded` as well: namely, I was able to get rid of `arrayTestNeeded` entirely, since it is equivalent to `typeTestsNeeded && arrayBlockNeeded`. So with that being the case, the five cases that `genCodeForUnsafeGetPut()` needs to handle are as follows (with case 3 being the one that I added just now):
```
   // There are 4 possible cases determining which checks and blocks should be added to the generated IL trees:
   // 1.) typeTestsNeeded && arrayBlockNeeded
   //     (i.e.: object type is unknown at compile time AND (offheap OR data element size < 4 (byte, short)))
   //     - tests: NULL test, Array test, Lowtag test
   //     - blocks: arrayDirectAccess, directAccess, indirectAccess
   // 2.) typeTestsNeeded && !arrayBlockNeeded && javaLangClass != NULL
   //     (i.e.: object type is unknown at compile time AND gencon AND data element size >= 4 (int, long) AND able to java/lang/Class)
   //     - tests: Lowtag test, NULL test, Class test
   //     - blocks: directAccess, indirectAccess
   // 3.) typeTestsNeeded && !arrayBlockNeeded && javaLangClass == NULL
   //     (i.e.: object type is unknown at compile time AND gencon AND data element size >= 4 (int, long) AND unable to java/lang/Class)
   //     - tests: NULL test, Lowtag test
   //     - blocks: directAccess, indirectAccess
   // 4.) !typeTestsNeeded && arrayBlockNeeded
   //     (i.e.: object is known to be array at compile time AND (offheap OR data element size < 4 (byte, short)))
   //     - tests: NULL test
   //     - blocks: arrayDirectAccess, directAccess
   // 5.) !typeTestsNeeded && !arrayBlockNeeded
   //     (i.e.: object is known to be non-array at compile time)
   //     - tests: none
   //     - blocks: directAccess
```
So now `javaLangClass == NULL` is separated out from the other conditions and only has an effect if `typeTestsNeeded == true` and `arrayBlockNeeded == false` (cases 2 and 3), which should cover the scenario that you mentioned above.

hzongaro replied:
> In the situation where gencon is being used and `conversionNeeded` is false, both sides of the array test diamond will lead to the `directAccessBlock`, so we can just skip the array test altogether. This leaves just the NULL and lowtag tests. . . .

I'm not sure that it's permissible to omit the array test in this case.  I haven't confirmed I'm correct about this, but I think it's possible that code could be writing or reading across array elements, even if conversion is not needed - for instance, I think it's possible that the caller could give an offset of 17 for an int array in a call to `Unsafe.putInt`, in order to write beginning at the last byte of one element and into the first three bytes of the next element.

Previously if `conversionNeeded` is false, the IL that was generated &mdash; for both cases where `javaLangClass` is null or is not null &mdash; allowed for the possibility that the low bit of the offset is set even if the location being stored to/loaded from is some Java object and not a static field.  The only way I can think that might be possible is with unaligned access across array elements - although `Unsafe` also has `Unaligned` methods that seem like they ought to be used in such situations.

Does anyone with more knowledge of how `Unsafe` works or is used understand why it needs to allow for the possibility that the low bit of the offset might be set for `{get|put}{Int|Long}` even if the target is not a static field?  @vijaysun-omr, @jdmpapin?

jdmpapin replied:
`get` and `put` for smaller integral types need to handle the case where a naturally aligned array element is not aligned to a 4-byte boundary. Is it possible that the logic to do so was needlessly applied for int and long as well?

The javadoc in `Unsafe` heavily restricts the valid offsets that can be passed. In the case of an array, the offset must be the array type's base offset plus a multiple of its index scale. Additionally, the array component type must be as expected by the method, e.g. for `getInt()` it must be `int`. Together these mean that anything passing a misaligned offset is violating the preconditions of the method. So *hopefully* nothing does that

Starting in Java 11 (though maybe introduced in 9 or 10) we have `Unsafe.getIntUnaligned()` and so on, which are there specifically to allow for potentially misaligned accesses. There is currently no special handling for those in the JIT compiler, but #19668 is open to treat them more or less just like their aligned counterparts. If the trees we generate for `getInt()` don't allow misaligned offsets, or if they stop allowing them, then the trees for `getIntUnaligned()` will need to differ in that respect (and similarly for `putInt()` and `get`/`put` of other types)

zl-wang replied:
i am wondering how directByteBuffer is accessed currently, whether it is through Unsafe APIs (obviously conversion must be prohibited, otherwise you can access over the buffer limit)?
if not, we might be able to get rid of the low-tagging shenanigans (J9 implementation-specific?).

hzongaro replied:
> Is it possible that the logic to do so was needlessly applied for int and long as well?

That's what I was wondering.  I'd been assuming these past few months that we were intentionally allowing for the possibility of unaligned access for int and long, but now I'm unsure about whether that was inadvertent.  I spent some time digging into the history, but it wasn't readily obvious whether that possibility was being explicitly allowed for.

However, I've just looked at the JVM's underlying implementation of `Unsafe.getLong()` in [`get64`](https://github.com/eclipse-openj9/openj9/blob/0e944e75c874f243f8f14c402ea8f268bf5e78b7/runtime/oti/UnsafeAPI.hpp#L262-L294).  It checks whether the object is an array before testing the least significant bit of the offset, and it's allowing for the possibility that an array access is misaligned.  So I think that the JIT's behaviour really was intended.

@tajila, am I understanding correctly?

tajila replied:
@hzongaro Yes, we do allow unaligned access with get/put* on arrays such that a read/load can cross an element boundary. Historically (pre-JDK9), this was always true. The VM implementation will always support this because we will have one version for pre and post JDK9. 

That being said,  recently tried this with an RI JDK17 build and it still allows unaligned access on an array.

> The javadoc in Unsafe heavily restricts the valid offsets that can be passed.

@jdmpapin where did you find this? it wasnt immediately clear to me.

jdmpapin replied:
Unsafe.java in the extension repos (sun/misc in JDK8, jdk/internal/misc in JDK11+). Look at the javadoc comment for `getInt()`, which is referenced from `putInt()` and other `get`/`put` methods

jdmpapin replied:
Oh, I guess we have our own Unsafe.java for 9+. But here's the [relevant part of the javadoc comment](https://github.com/ibmruntimes/openj9-openjdk-jdk8/blob/e642f8b86c560be492fa2d3c5a896d6ac5340021/jdk/src/share/classes/sun/misc/Unsafe.java#L107-L131) in Java 8

tajila replied:
Thanks, I was looking at the OpenJ9 docs. I wouldn't characterize "results are undefined" as a restrication, it is Unsafe after all.

Regardless, based on experimentation it seems that we still have to support unaligned access in get/putInt/Long at least on x86. There is no verification of parameters for Unsafe APIs so I suspect on some platforms it may just sigbus fault.

jdmpapin replied:
I interpreted "results are undefined" to mean that we hit undefined behaviour, i.e. that the implementation is theoretically free to assume that the arguments satisfy the stated conditions. I think generally for escape hatches like `Unsafe`, the possibility of arbitrary consequences for misuse (i.e. UB) is fundamentally what makes them "unsafe." If the comment just meant to say that the result/return value is unpredictable then it could have said that instead. AFAICT the comment is worded very carefully so that it applies to all of the regular `get`/`put` operations, which refer to it. Especially in the context of a `put`, I don't know how else to interpret "results are undefined"

But even on that interpretation, since we've been allowing a particular use, there might be code that relies on it. I guess your experimentation has shown that something is actually relying on it?

tajila replied:
Your interpretation of that comment sounds correct, but this would not be the first time an API defines certain pre-conditions yet the RI goes to great lengths to suppport cases outside those condition. 

Generally speaking, we treat "undefined behaviour" as spec'ed by the RI. Its hard to know who out there is relying on what behaviour. Given that unsafe is quite pervasive, I think the safest option is to aim for parity.

midronij replied:
Since unaligned array accesses are a possibility that I hadn't considered, I'll add back that array check for now. I'm not sure if there's another, more efficient way to account for this situation, but in the interest of getting these changes in soon, I think taking the safe approach here, and then adding any further improvements (if necessary) in a future contribution is probably the best course of action.

midronij replied:
With this change, the trees for the gencon/`javaLangClass == NULL` case can be viewed here: 
[gencon_int_unknown_javaLangClassNULL.log](https://github.com/user-attachments/files/16287454/gencon_int_unknown_javaLangClassNULL.log)

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1573,16 +1915,69 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   if (conversionNeeded && (arrayTestNeeded || arrayBlockNeeded))
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(callNodeWithConversion, conversionNeeded, true);
+   else if (arrayTestNeeded || arrayBlockNeeded)
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(callNodeTreeTop->getNode(), false, true);
```
As with `TR_J9InlinerPolicy::createUnsafePutWithOffset`, this might result in a block of IL that is identical to that produced for `directAccessTreeTop` above.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1294,15 +1414,51 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   //     iload        -> value to put
+
+   if (TR::Compiler->om.isOffHeapAllocationEnabled() && arrayBlockNeeded)
+   {
+      TR::Node *addrToAccessNode = arrayDirectAccessTreeTop->getNode()->getChild(0);
+
+      //change object base address to dataAddr
+      TR::Node *objBaseAddrNode = addrToAccessNode->getChild(0);
+      TR::Node *dataAddrNode = TR::TransformUtil::generateDataAddrLoadTrees(comp(), objBaseAddrNode);
+      addrToAccessNode->setChild(0, dataAddrNode);
+
+      //correct refcounts
+      objBaseAddrNode->decReferenceCount();
+      dataAddrNode->incReferenceCount();
+   }
```
The braces need to be indented to align with the code inside the block.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1399,31 +1555,145 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
-   TR::TreeTop *fallThroughTree = TR::TreeTop::create(comp(),callNodeTreeTop->getNode()->duplicateTree());
+#if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+   if (arrayTestNeeded)
+   {
```
The brace needs to be indented to align with the code inside the block.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1399,31 +1555,145 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
+      TR::Node *andNode = comp()->fej9()->testIsClassArrayType(vftLoad);
+      TR::Node *isArrayNode = TR::Node::createif(TR::ificmpne, andNode, TR::Node::create(TR::iconst, 0), NULL);
 
+      isArrayTreeTop = TR::TreeTop::create(comp(), isArrayNode, NULL, NULL);
+   }
+
+   if (arrayBlockNeeded)
+   {
```
The braces need to be indented to align with the code inside their respective blocks.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1399,31 +1555,145 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
+      //if array test is being generated, need to create non-array access treetop (same as default)
+      if (isArrayTreeTop)
+         nonArrayAccessTreeTop = TR::TreeTop::create(comp(),callNodeTreeTop->getNode()->duplicateTree());
+   }
```
The brace needs to be indented to align with the code inside the block.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1399,31 +1555,145 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
+
+   //only generate if and else trees for low tagged test in cases (1) and (3)
+   if (compareTree != NULL)
+   {
+      // genClassCheck generates a ifcmpne offset&mask 1, meaning if it is NOT
+      // lowtagged (ie offset&mask == 0), the branch will be taken
+      branchTargetTree = TR::TreeTop::create(comp(),callNodeTreeTop->getNode()->duplicateTree());
+      branchTargetTree->getNode()->getFirstChild()->setIsSafeForCGToFastPathUnsafeCall(true);
+
+      fallThroughTree = TR::TreeTop::create(comp(),callNodeTreeTop->getNode()->duplicateTree());
+
+
+      branchTargetTree->getNode()->getFirstChild()->setVisitCount(_inliner->getVisitCount());
+      fallThroughTree->getNode()->getFirstChild()->setVisitCount(_inliner->getVisitCount());
+
+      debugTrace(tracer(),"branchTargetTree = %p fallThroughTree = %p",branchTargetTree->getNode(),fallThroughTree->getNode());
+   }
```
The braces need to be indented to align with the code inside the block.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1440,30 +1710,53 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
 
-   callBlock->createConditionalBlocksBeforeTree(callNodeTreeTop,compareTree, branchTargetTree, fallThroughTree, comp()->getFlowGraph(),false,false);
+   if (arrayTestNeeded) //in case (3), we generate the array test diamond, followed by the low tagged check test
+   {
+      callBlock->createConditionalBlocksBeforeTree(callNodeTreeTop, isArrayTreeTop, arrayAccessTreeTop, nonArrayAccessTreeTop, comp()->getFlowGraph(), false, false);
+      nonArrayAccessTreeTop->getEnclosingBlock()->createConditionalBlocksBeforeTree(nonArrayAccessTreeTop, compareTree, branchTargetTree, fallThroughTree, comp()->getFlowGraph(), false, false);
+   }
+   else if (arrayBlockNeeded) //in case (2), no branching is needed: we simply need to replace the original CAS call with the modified array access block
+   {
+      callNodeTreeTop->insertAfter(arrayAccessTreeTop);
+      callNodeTreeTop->getPrevTreeTop()->join(callNodeTreeTop->getNextTreeTop());
+      callBlock->split(arrayAccessTreeTop->getNextTreeTop(), comp()->getFlowGraph(), true);
+      callBlock->split(arrayAccessTreeTop, comp()->getFlowGraph(), true);
+   }
```
The braces need to be indented to align with the code inside their respective blocks.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1440,30 +1710,53 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
-      TR::Node *branchTargetStoreNode = TR::Node::createWithSymRef(comp()->il.opCodeForDirectStore(dataType), 1, 1, branchTargetTree->getNode()->getFirstChild(), newSymbolReference);
-      TR::TreeTop *branchTargetStoreTree = TR::TreeTop::create(comp(), branchTargetStoreNode);
+      if (compareTree != NULL) //case (1) and (3) only
+      {
```
The brace needs to be indented.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1440,30 +1710,53 @@ TR_J9InlinerPolicy::createUnsafeCASCallDiamond( TR::TreeTop *callNodeTreeTop, TR
 
-      debugTrace(tracer(),"Inserted store tree %p for fall-through side of the diamond",fallThroughStoreNode);
+         debugTrace(tracer(),"Inserted store tree %p for fall-through side of the diamond", fallThroughStoreNode);
+      }
```
The brace needs to be indented.

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1573,16 +1915,69 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
+   //         lload    -> offset
+
+   if (TR::Compiler->om.isOffHeapAllocationEnabled() && arrayBlockNeeded)
+   {
+      TR::Node *addrToAccessNode;
+
+      if (conversionNeeded) //CASE 1
+         addrToAccessNode = arrayDirectAccessTreeTop->getNode()->getChild(0)->getChild(0)->getChild(0);
+      else //CASE 2
+         addrToAccessNode = arrayDirectAccessTreeTop->getNode()->getChild(0)->getChild(0);
+
+      //change object base address to dataAddr
+      TR::Node *objBaseAddrNode = addrToAccessNode->getChild(0);
+      TR::Node *dataAddrNode = TR::TransformUtil::generateDataAddrLoadTrees(comp(), objBaseAddrNode);
+      addrToAccessNode->setChild(0, dataAddrNode);
+
+      //correct refcounts
+      objBaseAddrNode->decReferenceCount();
+      dataAddrNode->incReferenceCount();
+   }
```
The braces need to be indented to align with the code in the block.

hzongaro reviewed (commented):
Most of my comments are pretty minor.  Even the one about duplicated blocks might not need to be addressed, or if it does, it might be that it could be handled in a follow on change. . . .

hzongaro commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -1494,6 +1787,56 @@ TR_J9InlinerPolicy::createUnsafeGetWithOffset(TR::ResolvedMethodSymbol *calleeSy
+
+   // If we are not able to get javaLangClass it is still inefficient to put direct Access far
+   // So in that case we will generate lowTagCmpTest to branch to indirect access if true
+   TR_OpaqueClassBlock *javaLangClass = comp()->fe()->getClassFromSignature("Ljava/lang/Class;",17, comp()->getCurrentMethod(),true);
```
It looks like I had removed it and replaced it with code in `genCodeForUnsafeGetPut`, but then you needed to revive for your changes.  Thanks for changing this to use `getClassClassPointer`!

hzongaro reviewed (commented):
I think the changes look good.  Just a couple of minor comments on some comments in the code.

Also, the Line endings check failed.  May I ask you to take a look at that?

Finally, I had a few comments on the commit comments.

- Looking at the description of commit e7441f266854fa3b865ecca22239d7a7119ad79d, it states that it's reverting commit 58efdd745783642b0a7d5265d51fe3c080182e96 - but looking at the two, it looks like the new commit is only reverting part of the x86 changes of that old commit.  May I ask you to update the commit comment to describe the change more explicitly?
- For commit 5608dae6b6d2aa7a2fd6ca2943dcd220ad58e9f4, per the [OpenJ9 Commit Guidelines](https://github.com/eclipse-openj9/openj9/blob/master/CONTRIBUTING.md#commit-guidelines), may I ask you to add a first line to the commit comment summarizing the commit?
- In the comment for commit 3a6a63fc5a7f0a294943747f2fee1534c4857742, a few of the lines are longer than 72 characters.

hzongaro commented:
One more late comment:  there is a [block of comments](https://github.com/eclipse-openj9/openj9/blob/5608dae6b6d2aa7a2fd6ca2943dcd220ad58e9f4/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L1251-L1330) before `createUnsafePutWithOffset` that describes the different ways that IL might be generated for Unsafe get and put operations.  May I ask you to update that to account for all the work you've done to add support for off-heap allocation?

midronij commented:
@hzongaro thank you for all of your feedback! I've made updates to address your most recent wave of suggested changes to both the comments in the code and the commit descriptions. Regarding the following:

> Looking at the description of commit https://github.com/eclipse-openj9/openj9/commit/e7441f266854fa3b865ecca22239d7a7119ad79d, it states that it's reverting commit https://github.com/eclipse-openj9/openj9/commit/58efdd745783642b0a7d5265d51fe3c080182e96 - but looking at the two, it looks like the new commit is only reverting part of the x86 changes of that old commit. May I ask you to update the commit comment to describe the change more explicitly?

Upon further inspection, I found that this commit was actually cancelled out by the very next one, https://github.com/eclipse-openj9/openj9/pull/17969/commits/16af9419b2b514f54e045699afe69e6072462cbf, so I simply dropped both of them.

If everything looks good, then I think we can move on to testing and merging! @zl-wang when you have a moment could you please get the CI tests started?

hzongaro reviewed (approved):
Looks good.  Thanks!

zl-wang commented:
Jenkins test sanity xlinux,aix jdk8,jdk21

pshipton commented:
Planning to revert this as the suspected cause of https://github.com/eclipse-openj9/openj9/issues/19947
